src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,380 @@
+            this.readContext = new ArrayList<>();
+        }
+
+        /**
+         * Careful: this ctor takes ownership of the passed reads and ReadContextData array.
+         * Do not modify them after this call (ideally don't even keep a reference to them).
+         */
+        private ContextShard(SimpleInterval interval, IntervalsSkipListOneContig<Variant> variants, final ArrayList<GATKRead> reads, final ArrayList<ReadContextData> readContext) {
+            this.interval = interval;
+            this.variants = variants;
+            this.reads = reads;
+            this.readContext = readContext;
+        }
+
+        /**
+         * create a new shard, keeping only the variants that overlap
+         * with the new interval. Reads etc are unchanged.
+         */
+        public ContextShard split(SimpleInterval newInterval) {
+            final IntervalsSkipListOneContig<Variant> newVariants;
+            if (null==variants) {
+                newVariants = null;
+            } else {
+                newVariants = new IntervalsSkipListOneContig<>( variants.getOverlapping(newInterval) );
+            }
+            return new ContextShard(newInterval, newVariants, reads, readContext);
+        }
+
+        /**
+         * creates a new shard, adding the specified variants.
+         * Note that readContext is unchanged (including the variants it may refer to).
+         */
+        public ContextShard withVariants(ArrayList<Variant> newVariants) {
+            return new ContextShard(this.interval, new IntervalsSkipListOneContig<>(newVariants), reads, readContext);
+        }
+
+        /**
+         * creates a new shard, adding the specified reads.
+         * Careful: this call takes ownership of the passed read array. So you can't modify it after this call.
+         */
+        public ContextShard withReads(ArrayList<GATKRead> newReads) {
+            return new ContextShard(this.interval, this.variants, newReads, readContext);
+        }
+
+        /**
+         * creates a new shard, adding the specified read context and *removing the variants*.
+         * Careful: this call takes ownership of the passed ReadContextData array. So you can't modify it after this call.
+         */
+        public ContextShard withReadContext(ArrayList<ReadContextData> newReadContext) {
+            return new ContextShard(interval, null, reads, newReadContext);
+        }
+
+        /**
+         * Returns the variants that overlap the query interval, in start-position order.
+         */
+        public ArrayList<Variant> variantsOverlapping(SimpleInterval interval) {
+            return variants.getOverlapping(interval);
+        }
+
+    }
+
+    /**
+     * Takes the variants, groups them into shards of size "bigShardSize" at the client, then
+     * runs subdivideAndFillReads then fillContext. See their documentation for more information.
+     */
+    public static PCollection<ContextShard> subdivideAndFillEverything(
```

Comment:
Can you name this top-level entry point something less awkward like AddContextDataToReadOptimized.add(), and supply an overload with fewer parameters that uses sensible defaults for bigShardSize, outputShardSize, margin, and marginForRef? We want this common operation of "adding overlapping data to a read" to be simple to use, if possible.